app/src/context/slices/helpers/pouchDBWrapper.ts

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2021, 2022 Macquarie University
+ *
+ * Licensed under the Apache License Version 2.0 (the, "License");
+ * you may not use, this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
+ * See, the License, for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Description:
+ *   Defines a wrapper around the PouchDB API so that we can catch
+ * and mitigate the indexedDB failure that occurs on IOS devices
+ */
+
+import {DatabaseInterface} from '@faims3/data-model';
+import {RUNNING_UNDER_TEST} from '../../../buildconfig';
+import {logError} from '../../../logging';
+import PouchDB from 'pouchdb-browser';
+
+// This is the PouchDB type which provides options for instantiating a database
+type LocalDatabaseOptions = PouchDB.Configuration.DatabaseConfiguration;
+
+// Default local options is none
+export const LOCAL_POUCH_OPTIONS: LocalDatabaseOptions = {};
+
+// enable memory adapter for testing
+if (RUNNING_UNDER_TEST) {
+  LOCAL_POUCH_OPTIONS['adapter'] = 'memory';
+}
+
+// Error types that indicate we should recreate the database connection
+const RECREATION_ERROR_NAMES = [
+  'InvalidStateError',
+  'AbortError',
+  'TransactionInactiveError',
+  'NotFoundError',
+  'DataError',
+  'VersionError',
+];
+
+export class PouchDBWrapper<T extends {}> implements DatabaseInterface<T> {
+  db: PouchDB.Database<T>;
+  name: string;
+  private recreationCount = 0;
+  private maxRecreations = 3;
+
+  constructor(name: string) {
+    // create a new database with this name and the standard options
+    this.name = name;
+    this.db = new PouchDB<T>(name, LOCAL_POUCH_OPTIONS);
+  }
+
+  private shouldRecreate(error: any): boolean {
+    // Check if this is an error type that indicates connection issues
+    return RECREATION_ERROR_NAMES.some(
+      errorName =>
+        error.name === errorName || error.message?.includes(errorName)
+    );
+  }
+
+  private async _recreate(): Promise<void> {
+    this.recreationCount++;
+    logError(
+      `PouchDB connection error, recreating (attempt ${this.recreationCount}) ${this.name}`
+    );
+
+    try {
+      // Destroy the old database instance
+      await this.db.close?.();
+    } catch (e) {
+      // Ignore errors when closing
+    }
+
+    // Create new instance
+    this.db = new PouchDB(this.name, LOCAL_POUCH_OPTIONS);
+  }
+
+  // Execute an operation with retry on certain errors
+  private async withRetry<R>(
+    operation: () => Promise<R> | R,
+    operationName: string
+  ): Promise<R> {
+    try {
+      return await operation();
```

Comment:
Are there any non-idempotent operations which can fail, but still side-effect the DB? 

If an error occurs you retry potentially multiple times - I see you only catch certain errors so hopefully this makes this impossible.